prov/verbs/src/windows/verbs_nd_init.c

```diff
@@ -0,0 +1,751 @@
+			       &((struct nd_adapter *)adapter2)->address);
+}
+
+static HRESULT nd_create_adapter(void)
+{
+	size_t addr_count = 0;
+	HRESULT hr;
+	struct nd_factory *factory = NULL;
+	struct nd_adapter *adapter = NULL;
+	ULONG listsize = 0;
+	wchar_t *saddr;
+	DWORD addrlen = 0;
+	UINT64 id;
+	int res;
+	ULONG linfo = sizeof(adapter->info);
+	ND2_ADAPTER_INFO *info = NULL;
+
+	for (size_t i = 0; i < nd_infra.class_factories.count; i++) {
+		factory = &nd_infra.class_factories.factory[i];
+		assert(factory->class_factory);
+
+		hr = factory->class_factory->lpVtbl->CreateInstance(
+			factory->class_factory, NULL, &IID_IND2Provider,
+			(void **)&factory->provider);
+		if (FAILED(hr))
+			return hr;
+		listsize = 0;
+		hr = factory->provider->lpVtbl->QueryAddressList(
+			factory->provider, NULL, &listsize);
+		if (hr != ND_BUFFER_OVERFLOW)
+			return hr;
+		if (!listsize) {
+			continue;
+		}
+
+		factory->addr_list = (SOCKET_ADDRESS_LIST *)malloc(listsize);
+		if (!factory->addr_list)
+			return ND_NO_MEMORY;
+
+		hr = factory->provider->lpVtbl->QueryAddressList(
+			factory->provider, factory->addr_list, &listsize);
+		if (FAILED(hr))
+			return hr;
+
+		for (size_t j = 0; j < factory->addr_list->iAddressCount; j++) {
+			if (ofi_nd_is_valid_addr(
+				    factory->addr_list->Address[j].lpSockaddr))
+				addr_count++;
+		}
+	}
+
+	if (!addr_count)
+		return E_NOINTERFACE;
+
+	nd_infra.adapters.adapter = (struct nd_adapter *)malloc(
+		addr_count * sizeof(*nd_infra.adapters.adapter));
+	if (!nd_infra.adapters.adapter)
+		return ND_NO_MEMORY;
+
+	/* put all available valid addresses into common array */
+	for (size_t i = 0; i < nd_infra.class_factories.count; i++) {
+		factory = &nd_infra.class_factories.factory[i];
+		for (size_t j = 0; j < factory->addr_list->iAddressCount; j++) {
+			if (ofi_nd_is_valid_addr(
+				    factory->addr_list->Address[j].lpSockaddr)) {
+				adapter = &nd_infra.adapters.adapter
+						   [nd_infra.adapters.count];
+				assert((int)sizeof(adapter->address) >=
+				       factory->addr_list->Address[j]
+					       .iSockaddrLength);
+				memcpy(&adapter->address,
+				       factory->addr_list->Address[j].lpSockaddr,
+				       factory->addr_list->Address[j]
+					       .iSockaddrLength);
+				adapter->factory = factory;
+				nd_infra.adapters.count++;
+			}
+		}
+	}
+
+	if (!nd_infra.adapters.count)
+		return E_NOINTERFACE;
+
+	/* sort adapters by addresses to set IP4 addresses first. then remove
+	   duplicates */
+	qsort(nd_infra.adapters.adapter, nd_infra.adapters.count,
+	      sizeof(struct nd_adapter), nd_adapter_cmp);
+	nd_infra.adapters.count =
+		unique(nd_infra.adapters.adapter, nd_infra.adapters.count,
+		       sizeof(struct nd_adapter), nd_adapter_cmp);
+
+	// Allocate memory for adapter_list, device_list, and context_list all at once.
+	nd_infra.adapters.adapter_list =
+		(struct nd_adapter **)calloc((nd_infra.adapters.count + 1) * 3,
+					     sizeof(nd_infra.adapters.adapter));
```

Comment:
We're mixing 3 different structures here and using a sizeof() value that isn't obviously related to any of them.  I think this should be something like:
```
calloc(nd_infra.adapters.count + 1,
    sizeof(*nd_infra.adapters.adapter_list) +
    sizeof(*nd_infra.adapters.device_list) +
    sizeof(*nd_info.adapters.context_list));
```